File: LeetCode/0946. Validate Stack Sequences/solution.go
/**
 * 151 / 151 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 3.6 MB 
 */
func validateStackSequences(pushed []int, popped []int) bool {
    s := []int{}
    i := 0
    for _, num := range pushed {
        s = append(s, num)
        for len(s) > 0 && s[len(s) - 1] == popped[i] {
            s = s[:len(s)-1]
            i++ 
        }
    }
    return len(s) == 0
}

File: LeetCode/0946. Validate Stack Sequences/Solution.java
/**
 * 151 / 151 test cases passed.
 * Runtime: 1 ms
 * Memory Usage: 41.1 MB 
 */
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        LinkedList<Integer> s = new LinkedList<>();
        int i = 0;
        for (int num : pushed) {
            s.addLast(num);
            while (!s.isEmpty() && s.getLast() == popped[i]) {
                s.pollLast();
                i += 1;
            }
        }
        return s.isEmpty();
    }
}
